### ISSUE TITLE
\underline and \boxaround do not honor stretcheability

### ISSUE THREAD

**Issue by Omikhleia** [sile-typesetter/sile#1320]
SILE 0.12.1

The `\underline` command and the (undocumented) `\boxaround` command from the **rules** package do not respect shrinking/stretching (caused by an explicit glue or even line justification)

```
\begin{document}
\script[src=packages/rules]
\language[main=und]% Do disable hyphenation and have stretched line easily
\nofolios

Some \underline{underlined text}

Some \boxaround{boxed text}

Some \underline{underlined\hfill text}\hfill

Some \boxaround{boxed\hfill text}\hfill

Some \underline{test lorem ipsum} dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 
Some ametloremipsum dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 

Some \boxaround{test lorem ipsum} dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 
Some ametloremipsum dor sit amet. Some test lorem ipsum dor sit amet. Some test lorem ipsum dor sit amet. 

\end{document}
```
![image](https://user-images.githubusercontent.com/18075640/155579328-a4720007-8ce6-46a0-a22f-37822992294f.png)

I got the problem with my own framebox package and I am attempting a fix for it - if it works there, I could possibly try to check the rules package.

By the way, the `boxaround` command is not documented, should we even keep it (it lacks options - perhaps I could propose my framebox package for replacement, let me know...)

**Comment by alerque** [comment]
I wondered how long it was going to take for this to come up.

I think it affects URL link areas too.

**Comment by Omikhleia** [comment]
> I think we need an API of some kind for wrapping content and then learning later what it's actual final shaped and placed size and location was and then being able to issue a callback rather than guessing based on the node shape information at the time the nodes are added.

Here, in this very kind of case, we do have that API, sort of.... In a box's `outputYourself()`, we have all the means to compute the actual size (and all things of that kind) with respect to the current line etc, 
Instead of the current underline/boxaround logic, which does a lot of re-boxing based on a wrong size (assumed out of context), size-dependent logic should likely be moved there.

In other terms, we do already have lot of (albeit not always well-documented) existing things in SILE, while a new API would call for possibly more problems. Moving framing logic there did the trick in my framebox stuff:

![image](https://user-images.githubusercontent.com/18075640/155627764-ea4c12d7-97a9-4deb-a956-7b12e46a054e.png)

=> So I would be almost ready, I guess, to tackle with the same kind of thing in the rules packages.

I do have one request for help, though... I'd possibly like to have a RTL version of my above example (possibly with the inner content colored, and a recommendation for a font to install), to test the RTL case. I have virtually no skill in reading those scripts, but I think some example would be neat. To keep it even simpler, some RTL test equivalent to this, and that could be used for command testing (just redefining the testcommand and checking the effect on the red text, even if one doesn't read the script and doesn't want to tackle with it):

```
\begin{document}
\script[src=packages/rules]
\script[src=packages/color]
\language[main=und]% disable hyphenation to have stretched line easily
\nofolios

Some \underline{\color[color=red]{underlined\hfill text}} test

\end{document}
```

**Comment by alerque** [comment]
We already have a couple fonts suitable for RTL languages in the test suite, I suggest Amiri as a generic one to use for such tests.

**Comment by Omikhleia** [comment]
> I think it affects URL link areas too.

I checked the code for `\pdf:link` and it seems legit (= it does the same kind of `outputYourself` logic that used in my PR). Or you were thinking to another command?

Regarding underlining, I failed to have it work properly in arabic : the underline is correctly placed, but the text content is off. Besides being ill-placed, I can't tell if the box content reversed or not (it does seem to be, but I can't read...)

![image](https://user-images.githubusercontent.com/18075640/155774978-66d170a1-4a76-44b7-94ce-9d9b258861ce.png)

The previous implementation suffered from the exact same type of issue on the same test file, so I'd say it's a separate issue...
